### PR TITLE
[MM-57080] Add Channel-specific notification sound options in new Notifications settings modal

### PR DESCRIPTION
**Pull request by M-ZubairAhmed**
#### Ticket Link
Jira https://mattermost.atlassian.net/browse/MM-57080

#### Screenshots
CleanShot 2024-07-22 at 21 48 14

#### Release Note

```release-note
Channel specific message notification sounds configuration added.
```

### CODE REVIEW

File: webapp/channels/src/components/widgets/modals/components/react_select_item.tsx [hunk 1]
```diff
@@ -9,12 +10,17 @@ import type {BaseSettingItemProps} from './base_setting_item';
 import BaseSettingItem from './base_setting_item';
 
 export type Option = {
-    value: number;
-    label: string;
+    value: string;
+    label: string | ReactNode;
```
Comment (devinbinnie) [inline webapp/channels/src/components/widgets/modals/components/react_select_item.tsx:14]: `ReactNode` includes `string`
```suggestion
    label: ReactNode;
```

**Review by devinbinnie — APPROVED** [review]

**Review by hmhealey — APPROVED** [review]

**Comment by M-ZubairAhmed** [comment]
@abhijit-singh can you please check out this one when you get time

**Review by abhijit-singh — CHANGES_REQUESTED** [review]
Looks good @M-ZubairAhmed. Noticed a couple of issues in the functionality and one minor cosmetic request:
1. For a new user, if they have never changed the `notification sound` setting in their account settings, then the notification sound setting under channel notification preferences does not have any sound selected. Ideally it should have `Bing` selected by default similar to the global setting. 

 Also, if you change it to some other sound at this point, the `Reset to default` button appears at the top of the section but it doesn't really do anything. 
 
  Note that both of these things work perfectly AFTER I change the global notification sound setting even once. 
  
  https://github.com/user-attachments/assets/58ab9edd-1cff-4286-b15d-a3b405a89516

2. Changing the sound in the channel notification preferences does not play it like it does in the global settings. Ideally, the sound should be played once when selected so that the user knows what it actually sounds like.

3. Can we reduce the width of the input box to ~224px to match with the design? Currently it looks too stretched out and it doesn't have to be.

**Review by yasserfaraazkhan — APPROVED** [review]
Just the 1st and 2nd points shared by Abhijeet.

The Sounds work in channel, GMs

**Comment by M-ZubairAhmed** [comment]
> Looks good @M-ZubairAhmed. Noticed a couple of issues in the functionality and one minor cosmetic request:
> 
>     1. For a new user, if they have never changed the `notification sound` setting in their account settings, then the notification sound setting under channel notification preferences does not have any sound selected. Ideally it should have `Bing` selected by default similar to the global setting.
> 
> 
> Also, if you change it to some other sound at this point, the `Reset to default` button appears at the top of the section but it doesn't really do anything.
> 
> Note that both of these things work perfectly AFTER I change the global notification sound setting even once.
> Screen.Recording.2024-07-30.at.11.29.47.AM.mov
> 
>     2. Changing the sound in the channel notification preferences does not play it like it does in the global settings. Ideally, the sound should be played once when selected so that the user knows what it actually sounds like.
> 
>     3. Can we reduce the width of the input box to ~224px to match with the design? Currently it looks too stretched out and it doesn't have to be.

@abhijit-singh all the points mentioned are now fixed

**Comment by abhijit-singh** [comment]
Thanks @M-ZubairAhmed. Those seem to be fixed but there's another thing that seems to be happening - 

When I did a fresh sign up of a new user, I could see the `Reset to default` button in Desktop Notifications under Channel Notification Preferences even though I had not changed anything from the default anywhere. Clicking on the button does not seem to do anything and the button does not go away. Ideally, the `Reset to default` button should not be there when the channel settings are matching with the global settings. 


https://github.com/user-attachments/assets/aa5cae4a-a732-4528-82be-b6f073963a9f

**Comment by M-ZubairAhmed** [comment]
> Thanks @M-ZubairAhmed. Those seem to be fixed but there's another thing that seems to be happening -
> 
> When I did a fresh sign up of a new user, I could see the `Reset to default` button in Desktop Notifications under Channel Notification Preferences even though I had not changed anything from the default anywhere. Clicking on the button does not seem to do anything and the button does not go away. Ideally, the `Reset to default` button should not be there when the channel settings are matching with the global settings.
> Screen.Recording.2024-08-05.at.4.38.00.PM.mov


I thought i fixed that but i didnt thanks for detailed analysis i fixed it now :)

**Comment by M-ZubairAhmed** [comment]
@yasserfaraazkhan adding you again since a lot of code change have happened after you previously approved

**Review by yasserfaraazkhan — APPROVED** [review]
Thank you @M-ZubairAhmed 
can you please check `channel_notifications_spec.js` e2e test case?

**Comment by abhijit-singh** [comment]
Thanks @M-ZubairAhmed. And I'm sorry but I found one more issue, not sure if it was not there before the most recent changes or I just missed it. Here's the issue:
- If I change the notification sound in my global notification setting, it does not get updated on existing channels and neither does it apply to a newly created channel. Hence, I still kept hearing the old notification sound for most messages.

https://github.com/user-attachments/assets/a54ce919-0d1a-4e8a-8a72-e1b24ac2693f

@yasserfaraazkhan Can you confirm if you see this issue?

**Comment by M-ZubairAhmed** [comment]
@abhijit-singh the issue mentioned above regarding the 

> If I never manually override my channel notification preferences OR when I click on Reset to default and Save, it should then keep my channel setting in sync with the global setting, until I manually change it again. The problem can be seen when the channel setting does not get updated even after changing the global setting.

Is opened into another ticket since this is same not only for channel notification sound but for other channel notification settings as well. Can you re verify in keeping above known bug

**Comment by M-ZubairAhmed** [comment]
> E2E test run is starting for commit `d4b20d47e9fd7a6dc0ba653547468a37bddb14d4`. You can check its progress by either:
> 
>     * Looking at the corresponding commit status, which will be available in a few moments: `E2E Tests/test`.
> 
>     * Looking at the [E2E test's Workflow Run](https://github.com/mattermost/mattermost/actions/runs/10390676789), with Run ID `10390676789`.
> 
>     * Looking at the [E2E test's Cycle on the Automation Dashboard](https://automation-dashboard.vercel.app/cycle/10390676789).

E2E shows no failures

**Review by abhijit-singh — APPROVED** [review]
Thanks @M-ZubairAhmed! Looks good to me apart from the `Reset to default` issue which we can tackle separately.

**Comment by Movion** [comment]
I can do the QA testing 👍

**Comment by lindy65** [comment]
Assigned to you for testig @Movion - let me know if you have any questions

**Comment by lindy65** [comment]
Hi @Movion - how's the QA review going?  Let me know if you have any questions :)

**Comment by Movion** [comment]
> Hi @Movion - how's the QA review going? Let me know if you have any questions :)

Heo @lindy65, I have tested all the options I know of and everything looks good.
As the number of channels and settings increases, it could quickly become confusing, which is why you should consider creating some kind of "overview". But that is not something that is important for the current PR.

The functions as presented work without any problems.

**Review by lindy65 — APPROVED** [review]
Thanks @M-ZubairAhmed - I have also reviewed the PR, taking comments into consideration and, apart from the "Reset to default" issue that is being tracked in a separate issue, everything looks good to me 👍

**Comment by lindy65** [comment]
Awesome @Movion! Thanks for your contribution and testing. Would love your help on future PR QA review requests :)

**Comment by lindy65** [comment]
>As the number of channels and settings increases, it could quickly become confusing, which is why you should consider creating some kind of "overview".

@M-ZubairAhmed - will you create a ticket to consider `@movion's` suggestion for this improvement in a later iteration?

**Comment by amyblais** [comment]
@M-ZubairAhmed Please help open a manual cherry-pick PR for ``release-9.5`` branch.

**Comment by M-ZubairAhmed** [comment]
@Movion @lindy65 

Can you please explain more about what you mean by 

> As the number of channels and settings increases, it could quickly become confusing, which is why you should consider creating some kind of "overview"

**Comment by Movion** [comment]
@M-ZubairAhmed 
It could happen that after a certain time many channels with different sounds are set. Currently there is only the possibility to check the setting directly in the channels. My suggestion is to create an overview page of the channels and their settings to counteract the complexity. 
Whether this makes sense or not is up to the product owners. It's just an idea that came to me while I was testing.

**Comment by lindy65** [comment]
@M-ZubairAhmed - I agree with `@movion` that having a modal that lists notification sounds for all channels as well as the default notification sound that's set in Settings > Notifications > Desktop notification sounds could be a nice improvement in a future iteration.  Perhaps also being possible to change the notification sound for each channel from that modal.

**Comment by M-ZubairAhmed** [comment]
@Movion would you be open for creating a github issue with your idea suggestion?

**Pull request merged**